eth/protocols/snap/sync.go

```diff
@@ -1910,13 +1974,29 @@ func (s *Syncer) processAccountResponse(res *accountResponse) {
 			}
 		}
 	}
-	// Delete any subtasks that have been aborted but not resumed. This may undo
-	// some progress if a new peer gives us less accounts than an old one, but for
-	// now we have to live with that.
-	for hash := range res.task.SubTasks {
-		if _, ok := resumed[hash]; !ok {
-			log.Debug("Aborting suspended storage retrieval", "account", hash)
-			delete(res.task.SubTasks, hash)
+	// Delete any subtasks that have been aborted but not resumed. It's essential
+	// as the corresponding contract might be self-destructed in this cycle(it's
+	// no longer possible in ethereum as self-destruction is disabled in Cancun
+	// Fork, but the condition is still necessary for other networks).
+	//
+	// Keep the leftover storage tasks if they are not covered by the responded
+	// account range which should be picked up in next account wave.
+	if len(res.hashes) > 0 {
+		// The hash of last delivered account in the response
+		last := res.hashes[len(res.hashes)-1]
+		for hash := range res.task.SubTasks {
+			// TODO(rjl493456442) degrade the log level before merging.
+			if hash.Cmp(last) > 0 {
+				log.Info("Keeping suspended storage retrieval", "account", hash)
+				continue
```

Comment:
I'm a bit torn about this one.

The issue is that the large contracts *covered* by the newly delivered account range will have their root hash updated, so retrieving chunks from them will be possible. However, the root hash of contracts *not covered* by the newly delivered account range will have their old root hash.

If the pivot moves and a contract was updated with the move but not covered by the new account range, then no storage request will be fulfillable, because the storage root will be different. That would I think lead to a subtask never being able to finish, forever stalling the snap sync.

Reply:
IMO if we want to retain started-but-not-yet-finished subtasks that fall *outside* of the delivered account range, we'd need to introduce some form of suspension to disregard them when moving the account range forward (and I'm guessing also not start state sync tasks for them).

Reply:
As @rjl493456442 pointed out, this is fine because the effectiveSubTasks will only start up those tasks that are covered by the account range (and hence have correct state roots).

https://github.com/ethereum/go-ethereum/pull/29378/files#r1557348963